### DISCUSSION TITLE
What is the scope of the codeql queries?

### DISCUSSION THREAD

**Discussion by kakashiUc** [github/codeql#5649] [Q&A] [answered]
I have 5 files in a directory: - 

file1.py
file2.py
module1/file3.py
module2/file4.py
module2/file5.py

There are three ways in which I can create databases and analyze them.

1. One way is to create databases of the files independently and run the same query suite on each of these separate databases to analyze them. 
File structure would be single files, i.e, file2.py or module1/file3.py

2. Second way is to create a database with all the five files together and run the query suite on the database; with the actual directory structure.

3. Third way is to create a database with all the five files together and run the query suite on the database; but with all files in a single directory. In this case -
file1.py
file2.py
file3.py
file4.py
file5.py

Is there a possibility of the results being different in these three cases?
Does the database creation process and subsequently the queries consider information from other files present in the same directory, other than the current file?

**Comment by criemen** [comment, marked as answer]
Depending on your code, all three ways of building the database will yield different analysis results.
CodeQLs strength lies in its interprocedural analyses - we do consider all your code, in all your files.
In option 1, you will give up many possibilities for us to find bugs in your code.

You should analyse your project in the structure that it's written, so that we can see all the code.
This also saves you the work of re-arranging your project.
I hope that helps, please let me know if you have any more questions.

**Reply by kakashiUc** [reply]
Thanks a lot! That does answer my question.
I have one more question. Is it possible to know which other program files and which lines of code were used to answer a query when analyzing a program file? In short what context was used to answer a certain query.

**Reply by criemen** [reply]
No, that's not possible.
The analysis never has a per-file view of your code, it always has access to your entire project.

Some queries then only look at one class, or one function, whereas some analyses like dataflow work across your entire project.
This distinction is not made explicit anywhere, to my knowledge.

However, if you have a dataflow/taint-tracking based query, for example for a SQL injection, we'll display an example path through your program showing how you can get the sql injection. This shows, for these specific queries, the context in your program that was used by the analysis.

**Reply by kakashiUc** [reply]
Thanks a lot!
